// @grafbase/sdk/dist/src/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = void 0;
var auth_1 = require("./auth");
var cache_1 = require("./cache");
var experimental_1 = require("./experimental");
var federation_1 = require("./federation");
/**
 * Defines the complete Grafbase configuration.
 */
var Config = /** @class */ (function () {
    function Config(input) {
        this.schema = input.schema;
        if (input.auth) {
            this.auth = new auth_1.Authentication(input.auth);
        }
        if (input.cache) {
            this.cache = new cache_1.GlobalCache(input.cache);
        }
        if (input.experimental) {
            this.experimental = new experimental_1.Experimental(input.experimental);
        }
        if (input.federation) {
            this.federation = new federation_1.Federation(input.federation);
        }
    }
    Config.prototype.toString = function () {
        var schema = this.schema.toString();
        var auth = this.auth ? this.auth.toString() : '';
        var cache = this.cache ? this.cache.toString() : '';
        var experimental = this.experimental ? this.experimental.toString() : '';
        var federation = this.federation ? this.federation.toString() : '';
        return "".concat(experimental).concat(auth).concat(cache).concat(federation).concat(schema);
    };
    return Config;
}());
exports.Config = Config;
